### PR TITLE
Save training labels of classifiers

### PR DESCRIPTION
**Pull request by jluethi**
**Adding new functionality to the viewer & the classifier tool**
This pull request adds the functionality to save the selections made in the viewer during training of classifiers (SVC). It adds the option to the web interface to save such selections when training a classifier and is on by default, because saving of the training input should be done when possible to be able to reproduce results.
It also adds the option to assign a name to the classifier & the saved labels before submitting a job to allow users to maintain the overview over what result is related to what.

**Installation on other instances**
As there are changes both to the library code & the web interface code, it is necessary to restart the `tm_server` and to run `gulp --prod` on `tmui` to recompile the typescript code. It can be done in the following way:
```
sudo supervisorctl restart tm_server

export PATH=/srv/tissuemaps/pelkmanslab-tm-frontend-YOURNAME/src/tmui/src/node_modules/.bin:$PATH
cd /srv/tissuemaps/pelkmanslab-tm-frontend-YOURNAME/src/tmui/src
gulp --prod
```

**Implementation details**
I first intended to add this as a separate tool to the interface. As the addition of tools is undocumented, I could not figure out how to successfully make a tool window appear in the interface (I got to the button for it, but had error messages I couldn't trace afterwards). As labels are typically assigned for the SVC tool anyway, I decided to integrate it into that tool.
The implementation as part of SVC is reasonable, because it allows to set the default that the training data for a classifier is always saved, which is a desirable target.
I also added the option in the interface to toggle either training the classifier or saving the labels off, if a user only wants to perform one action. This is handled in angular.js code and only the necessary jobs are submitted to tm_server.
Within the tm_library, I had to restructure the process handling in the classifier. Simpler approaches that did not rely on two separate submissions hit limits within TissueMaps's design. If only one submission is made, only one location for saving results can be registered, thus the existing saving of results would not work to save both labels & classifier results. I refactored the code such that two submissions are triggered when both options are selected in the interface (default). This has two minor downsides in the interface:
1. The user gets two confirmation of successful job submissions
2. The saved labels & the classifier have different job ids, thus making the pairing non-obvious (but they always follow each other and the user can now name them the same way)

Those downsides are minor though and any other implementation would required large-scale refactoring of TissueMaps workflows.
I also improve the handling of the viewer regarding unspecified labels when displaying classifier results. Those now appear as black cells instead of throwing errors.

### CODE REVIEW

File: tmlibrary/tmlib/tools/base.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/base.py]: These could become:
```suggestion
                name = result_attributes.get('name', self.__class__.__name__)
```
if `result_attributes` is a "classic" dictionary, see: https://docs.python.org/3/library/stdtypes.html#dict.get

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]:
```suggestion
            raise ValueError('Tool {} is not implemented'.format(payload['task']))
```

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]:
```suggestion
            submission_id, mapobject_type_name,
            result_type='SupervisedClassifierToolResult', name=result_name,
            unique_labels=unique_labels, label_map=label_map
```

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]: Is this still needed? In case you want to keep it, explain better what kind of "result" are you talking about (think about your sysadmins scrolling through the logs :smile:)

Is this a really important line that deserves a "info" status in the logs? Log lines are divided into many "levels", usually DEBUG, INFO, WARNING, ERROR. I'd set it as a debug line.

If you keep it, make it a single line, otherwise there may be other lines that end up in the middle, sth like:
```suggestion
        logger.debug('Result id from the classification: ' + str(result_id))
```

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]: See comment above:
```suggestion
        logger.debug('Saving current selections for submission id: ' + str(submission_id))
```

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]:
```suggestion
        for cls_id, cls in enumerate(payload['training_classes']):
            labels.update({val: float(cls_id) for val in cls['object_ids']})
            label_map[float(cls_id)] = {'name': cls['name'], 'color': cls['color']}
```

I'd avoid single-letter variables as much as I can.

Question: are you sure that there are no "overlapping" values in `cls['objects_id']` that could end up as the same key of those dictionaries?

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]:
```suggestion
        # Hard-coding tpoint as 0 is suboptimal, but I don't know how I
        # would get the actual tpoint information and as far as I can see, it
        # does not matter. It's just required to create the same kind of data
        # structure to be saved as in perform_classification.
        # To fix this, check how it's done in load_feature_values
        tpoint = 0
        indices = [(label, tpoint) for label in labels.keys()]
```

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]:
```suggestion
```
You did already define it at line 214.
Moreover, as `label_array` isn't modified, probably using np.unique on it is faster.

If unique_labels is not used as an array but e.g. is used as a list and you only need to filter duplicates, you could use unique_labels = set(labels.values()); if the labels are a lot, let's say >1000, np.unique could still be faster, as numpy probably implements it as a vectorized function.

File: tmlibrary/tmlib/tools/classification.py
Comment (bebosudo) [inline tmlibrary/tmlib/tools/classification.py]: Same comment as above.

File: tmui/src/app/src/core/tool/LabelLayer.ts
Comment (bebosudo) [inline tmui/src/app/src/core/tool/LabelLayer.ts]:
```suggestion
            if (label === "None") {
```

I'm no js/typescript expert, but probably here a triple-equal check is useful: https://stackoverflow.com/questions/359494/

File: tmui/src/app/src/core/tool/LabelLayer.ts
Comment (bebosudo) [inline tmui/src/app/src/core/tool/LabelLayer.ts]: Probably here you can use the static value defined here:
https://github.com/pelkmanslab/TissueMAPS/blob/071f3cfd0ff9f52af7d18645ba59bb1b00fd6dcb/tmui/src/app/src/core/util/Color.ts#L161-L163
and set it as "default": this probably could remove the gulp error at compile time.
If then the value is undefined, use the label to extract the color.

```suggestion
             var fillColor: ol.Color.BLACK;
             if (label === undefined) {
                fillColor = this.colorMapper(label).toOlColor();
             } 
```
(I wasn't able to create a multiline patch so you may need to edit it manually)

File: tmui/src/app/src/core/tool/result/SupervisedClassifierLabelLayer.ts
Comment (bebosudo) [inline tmui/src/app/src/core/tool/result/SupervisedClassifierLabelLayer.ts]: Fix indentation:
```suggestion
            colorMap[label] = Color.fromHex(this.attributes.label_map[label]['color']);
```

**Review by bebosudo — CHANGES_REQUESTED** [review]

**Comment by bebosudo** [comment]
Joel, hope I didn't overwhelm you with comments :)

File: tmlibrary/tmlib/tools/classification.py
Comment (jluethi) [reply tmlibrary/tmlib/tools/classification.py]: Ah, one of my many logging entries made it through. No, that was debugging online. Thanks for catching this :)

File: tmlibrary/tmlib/tools/classification.py
Comment (jluethi) [reply tmlibrary/tmlib/tools/classification.py]: @bebosudo This isn't actually new code, but just a copy of how the classifier extracted the object_ids from the payload. Given that it worked robustly there, I assume that's not an issue, but I don't have a good setup to test that

File: tmlibrary/tmlib/tools/classification.py
Comment (jluethi) [reply tmlibrary/tmlib/tools/classification.py]: Another unnecessary log, was just for debugging :)

File: tmui/src/app/src/core/tool/LabelLayer.ts
Comment (jluethi) [reply tmui/src/app/src/core/tool/LabelLayer.ts]: I tried this, but when I use `var fillColor: ol.Color.BLACK;,` I get `app/src/core/tool/LabelLayer.ts(79,31): error TS2305: Module 'ol' has no exported member 'Color'.`
I don't know what this 'ol' is to be honest or where it is defined...
I also tried directly using `fillColor = new Color(0, 0, 0);`, but then I get `app/src/core/tool/LabelLayer.ts(82,17): error TS2322: Type 'Color' is not assignable to type 'ol.Color'.` So we seem to have different kind of color classes here...

Also, unfortunately, I can't check for undefined. Took me half an hour to figure this out, but it's actually the exact string "None", not a null or other undefined value that is given for undefined values here.

Comment (jluethi) [reply tmui/src/app/src/core/tool/LabelLayer.ts]: For reference, with my original code here, I get `app/src/core/tool/LabelLayer.ts(84,17): error TS2322: Type 'string' is not assignable to type 'Color'.`

**Comment by jluethi** [comment]
@bebosudo Thanks a lot for the review! I've made all the suggested changes except for the one with the black color. That one seems more complicated, see comments above. The rest is changed and I just tested the changes, everything still runs fine 👍

File: tmui/src/app/src/core/tool/LabelLayer.ts
Comment (bebosudo) [reply tmui/src/app/src/core/tool/LabelLayer.ts]: @jluethi and what if you try `Color.BLACK.toOlColor()` instead of `ol.Color.BLACK` as I initially suggested?

**Review by bebosudo — COMMENTED** [review]
Check Color libraries

File: tmui/src/app/src/core/tool/LabelLayer.ts
Comment (jluethi) [reply tmui/src/app/src/core/tool/LabelLayer.ts]: That worked!! :) Thanks @bebosudo !

**Comment by jluethi** [comment]
@bebosudo Ok, the color part is fixed now as well. Now it runs without a compilation error and still works nicely in the user interface afterwards. Thus, all the requested changes were applied. I suggest we wait for @scottberry review and then it should be ready to be merged.

File: tmlibrary/tmlib/tools/classification.py
Comment (scottberry) [inline tmlibrary/tmlib/tools/classification.py]: for consistency I don't think camelCase is used at all in TissueMAPS, Therefore `saveLabels` should be `save_labels` throughout

File: tmlibrary/tmlib/tools/classification.py
Comment (scottberry) [inline tmlibrary/tmlib/tools/classification.py:238]: will this lead to errors in database if names are not unique?

Comment (jluethi) [reply tmlibrary/tmlib/tools/classification.py:238]: No, the actual names always include the submission id and tissuemaps ensures that this is unique. So far, everything was just named "Classification-19" for id 19 etc. Now it's a user-defined name +"-19" for submission id 19

Comment (jluethi) [reply tmlibrary/tmlib/tools/classification.py:238]: I've tested dozens of rounds with varying names, names that are too long, duplicate names etc. and it runs stably

**Review by scottberry — DISMISSED** [review]
This is awesome! I can't believe you finally did this after...hmmm, 3 years! well done. I am not familiar with the classification code but can't see anything obviously wrong.

File: tmlibrary/tmlib/tools/classification.py
Comment (scottberry) [reply tmlibrary/tmlib/tools/classification.py:238]: Cool, perhaps you could add a warning "(max. 20 chars)" above the box where you write the name

Comment (jluethi) [reply tmlibrary/tmlib/tools/classification.py:238]: Yes, I had already added that :)

**Review by bebosudo — APPROVED** [review]

**Comment by bebosudo** [comment]
LGTM, we can merge this if you're ok with the last edit @scottberry.

**Review by scottberry — APPROVED** [review]

**Comment by jluethi** [comment]
@scottberry Thanks for your review! I renamed saveLabels to save_labels as suggested. The rest was already covered, see comments above.
We can merge into master now

**Pull request merged**
